src/wallet/wallet.cpp

```diff
@@ -95,79 +95,116 @@ CPubKey CWallet::GenerateNewKey(CWalletDB &walletdb, bool internal)
-    CExtKey childKey;              //key at m/0'/0'/<n>'
-
     // try to get the master key
-    if (!GetKey(hdChain.masterKeyID, key))
+    if (GetKey(hdChain.masterKeyID, key))
+    {
+        // for now we use a fixed keypath scheme of m/0'/0'/k
+        CExtKey masterKey;             //hd master key
+        CExtKey accountKey;            //key at m/0'
+        CExtKey chainChildKey;         //key at m/0'/0' (external) or m/0'/1' (internal)
+        CExtKey childKey;              //key at m/0'/0'/<n>'
+
+        masterKey.SetMaster(key.begin(), key.size());
+
+        // derive m/0'
+        // use hardened derivation (child keys >= 0x80000000 are hardened after bip32)
+        masterKey.Derive(accountKey, BIP32_HARDENED_KEY_LIMIT);
+
+        // derive m/0'/0' (external chain) OR m/0'/1' (internal chain)
+        assert(internal ? CanSupportFeature(FEATURE_HD_SPLIT) : true);
+        accountKey.Derive(chainChildKey, BIP32_HARDENED_KEY_LIMIT + (internal ? 1 : 0));
+
+        // derive child key at next index, skip keys already known to the wallet
+        do {
+            // always derive hardened keys
+            // childIndex | BIP32_HARDENED_KEY_LIMIT = derive childIndex in hardened child-index-range
+            // example: 1 | BIP32_HARDENED_KEY_LIMIT == 0x80000001 == 2147483649
+            if (internal) {
+                chainChildKey.Derive(childKey, hdChain.nInternalChainCounter | BIP32_HARDENED_KEY_LIMIT);
+                metadata.hdKeypath = "m/0'/1'/" + std::to_string(hdChain.nInternalChainCounter) + "'";
+                hdChain.nInternalChainCounter++;
+            }
+            else {
+                chainChildKey.Derive(childKey, hdChain.nExternalChainCounter | BIP32_HARDENED_KEY_LIMIT);
+                metadata.hdKeypath = "m/0'/0'/" + std::to_string(hdChain.nExternalChainCounter) + "'";
+                hdChain.nExternalChainCounter++;
+            }
+        } while (HaveKey(childKey.key.GetPubKey().GetID()));
+        secret = childKey.key;
+        pubkey = childKey.key.GetPubKey();
+        metadata.hdMasterKeyID = hdChain.masterKeyID;
+        // update the chain model in the database
+        if (!walletdb.WriteHDChain(hdChain))
+            throw std::runtime_error(std::string(__func__) + ": Writing HD chain model failed");
+    }
+    else if (IsExternalHD()) {
+        CExtPubKey& masterKey = hdChain.externalHD;             //hd master key
+        CExtPubKey chainChildKey;         //key at m/0 (external) or m/1 (internal)
+        CExtPubKey childKey;              //key at m/0/<n>
+
+        // derive m/x
+        masterKey.Derive(chainChildKey, internal ? 1 : 0);
+
+        // derive child key at next index, skip keys already known to the wallet
+        do {
+            if (internal) {
+                chainChildKey.Derive(childKey, hdChain.nInternalChainCounter);
+                metadata.hdKeypath = "m/1/" + std::to_string(hdChain.nInternalChainCounter);
```

Comment:
for other reviewers: the account key level is dropped here because it's hardened in BIP44. Maybe a source code comment would be good.